example_workspace/src/image_transport_sample_app/jni/CMakeLists.txt

```diff
@@ -0,0 +1,61 @@
+#
+# Copyright (C) The Android Open Source Project
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+cmake_minimum_required(VERSION 3.4.1)
+project(image_transport_sample)
+
+find_package(catkin REQUIRED COMPONENTS roscpp std_msgs rosconsole image_transport compressed_image_transport theora_image_transport)
+
+include_directories(${catkin_INCLUDE_DIRS})
+
+# build native_app_glue as a static lib
+set(${CMAKE_C_FLAGS}, "${CMAKE_C_FLAGS}")
+add_library(native_app_glue STATIC
+    ${ANDROID_NDK}/sources/android/native_app_glue/android_native_app_glue.c)
+
+# now build app's shared lib
+set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=gnu++11 -Wall -Werror")
+
+# Export ANativeActivity_onCreate(),
+# Refer to: https://github.com/android-ndk/ndk/issues/381.
+set(CMAKE_SHARED_LINKER_FLAGS
+    "${CMAKE_SHARED_LINKER_FLAGS} -u ANativeActivity_onCreate")
+
+add_library(${PROJECT_NAME} SHARED 
+    src/main/cpp/image_transport_plugin_test.cpp
+)
+
+target_include_directories(${PROJECT_NAME} PRIVATE
+    ${ANDROID_NDK}/sources/android/native_app_glue)
+
+# This is to be sure that the symbols of this libraries aren't dropped
+# Also, put them in catkin find_package, in order to solve transitive dependencies without problems.
+SET (WHOLE_ARCHIVE_LIBRARIES
+     -Wl,--whole-archive
+     $ENV{TARGET_DIR}/lib/libimage_transport_plugins.a
+     $ENV{TARGET_DIR}/lib/libtheora_image_transport.a
+     $ENV{TARGET_DIR}/lib/libcompressed_image_transport.a
+     -Wl,--no-whole-archive
+)
+
+# add lib dependencies
+target_link_libraries(${PROJECT_NAME}
+    android
+    native_app_glue
+    log
+    ${WHOLE_ARCHIVE_LIBRARIES}
+    ${catkin_LIBRARIES}
```

Comment:
Yes, here link order is important. Maybe another solutions is to patch all the plugins to export also the flags (-Wl,--whole-static library_name -Wl,--no-whole-archive) for android in catkin_package LIBRARIES option (I think this is possible with #68). Maybe is more difficult to maintain (I don't think it will be accepted upstream), but easier to use for the user.
@meyerj WDYT?

Reply:
1. Patching plugin packages only works for the packages in the `catkin_ws`, but is not a general solution for packages providing plugins in other workspace(s) listed in `--extends-workspace`.

2. `catkin_LIBRARIES` deduplicates libraries found in all packages and does not preserve the order of listed libraries: [catkinConfig.cmake.in:97-102](https://github.com/ros/catkin/blob/kinetic-devel/cmake/catkinConfig.cmake.in#L97-L102)

   What might work is to add the string `"-Wl,--whole-static library_name -Wl,--no-whole-archive"` to `catkin_LIBRARIES`, with spaces and not as three separate list entries. Should check whether `target_link_libraries(foo ${catkin_LIBRARIES})` would accept this.

We were thinking about adding a CMake macro `pluginlib_export_plugins()` in a [catkin CFG_EXTRAS](http://docs.ros.org/jade/api/catkin/html/dev_guide/generated_cmake_api.html#catkin-package) script provided by `pluginlib`, or a separate package. The macro needs to be called before `catkin_package()` in packages providing plugins and generates another cmake script which appends to a list of plugins and sets another variable, e.g. `pluginlib_EXPORTED_PLUGIN_LINKER_FLAGS`, which then in turn is added to `${PROJECT_NAME}_CFG_EXTRAS` again. Like that, each package can statically link to all plugins provided by all its dependencies by adding those flags to `target_link_libraries(...)`, without being affected by catkin internals. It is similar to the mechanics of `genmsg` and friends.

On the downside, linking to all plugins with `--whole-archive` might blow up binary sizes.

Maybe it is not even necessary to patch packages that provide macros if the cmake script provided by `pluginlib` already parses `package.xml` and the exported plugin declarations and finds plugins on it own.

On the long term exporting plugins like this might even replace the [pluginlib_helper in ros_android](https://github.com/Intermodalics/ros_android/tree/kinetic/files/pluginlib_helper), because the same set of CMake scripts could also generate a source file with the descriptions of all listed plugins on the fly - similar to `pluginlib_helper.cpp` - and compile and link it to the target with a single `pluginlib_link_plugins(target [plugin1 [plugin2 ...])` call that needs to be added by the user in the Android-specific `CMakeLists.txt` for his application.